oss-internship-2020/guetzli/guetzli_entry_points.cc

```diff
@@ -0,0 +1,309 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "guetzli_entry_points.h"
+
+#include <sys/stat.h>
+
+#include <algorithm>
+#include <cstdio>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "guetzli/jpeg_data_reader.h"
+#include "guetzli/quality.h"
+#include "png.h"
+#include "sandboxed_api/sandbox2/util/fileops.h"
+#include "sandboxed_api/util/statusor.h"
+
+namespace {
+
+constexpr int kBytesPerPixel = 350;
+constexpr int kLowestMemusageMB = 100;
+
+struct GuetzliInitData {
+  std::string in_data;
+  guetzli::Params params;
+  guetzli::ProcessStats stats;
+};
+
+struct ImageData {
+  int xsize;
+  int ysize;
+  std::vector<uint8_t> rgb;
+};
+
+sapi::LenValStruct CreateLenValFromData(const void* data, size_t size) {
+  void* new_data = malloc(size);
+  memcpy(new_data, data, size);
+  return { size, new_data };
+}
+
+sapi::StatusOr<std::string> ReadFromFd(int fd) {
+  struct stat file_data;
+  int status = fstat(fd, &file_data);
+  
+  if (status < 0) {
+    return absl::FailedPreconditionError(
+      "Error reading input from fd"
+    );
+  }
+  
+  std::string result;
+  result.resize(file_data.st_size);
+  status = read(fd, result.data(), result.size());
+
+  if (status < 0) {
+    return absl::FailedPreconditionError(
+      "Error reading input from fd"
+    );
+  }
+
+  return result;
+}
+
+sapi::StatusOr<GuetzliInitData> PrepareDataForProcessing(
+                                const ProcessingParams* processing_params) {
```

Comment:
this is not part of the C interface so const reference instead of const pointer can be used and is preferred.